core-processor/src/processor.rs

```diff
@@ -233,87 +231,192 @@ pub fn precharge(
     }
 }
 
-/// Prepares environment for the execution of a program.
-/// Checks if there is a required export and tries to charge for code and memory pages.
-/// Returns either `PreparedMessageExecutionContext` for `process` or an array of journal notes.
-/// See `PrepareResult` for details.
-pub fn prepare(
+/// Charge a message for fetching the length of the binary code.
+pub fn precharge_for_code_length(
     block_config: &BlockConfig,
-    execution_context: MessageExecutionContext,
-) -> PrepareResult {
+    dispatch: PrechargedDispatch,
+    destination_id: ProgramId,
+    executable_data: Option<ExecutableActorData>,
+) -> PrechargeResult<ContextChargedForCodeLength> {
     use executor::ChargeForBytesResult;
 
-    let read_per_byte_cost = block_config.read_per_byte_cost;
     let read_cost = block_config.read_cost;
-    let max_reservations = block_config.max_reservations;
-
-    let MessageExecutionContext {
-        actor,
-        precharged_dispatch,
-        origin,
-        subsequent_execution,
-    } = execution_context;
-    let Actor {
-        balance,
-        destination_program: program_id,
-        executable_data,
-    } = actor;
 
-    let (dispatch, mut gas_counter, mut gas_allowance_counter) = precharged_dispatch.into_parts();
+    let (dispatch, mut gas_counter, mut gas_allowance_counter) = dispatch.into_parts();
 
     let actor_data = match check_is_executable(executable_data, &dispatch) {
         Err(status_code) => {
-            return PrepareResult::Error(process_non_executable(dispatch, program_id, status_code));
+            return Err(process_non_executable(
+                dispatch,
+                destination_id,
+                status_code,
+            ));
         }
         Ok(data) => data,
     };
 
     if !actor_data.code_exports.contains(&dispatch.kind()) {
-        return PrepareResult::WontExecute(process_success(
+        return Err(process_success(
             SuccessfulDispatchResultKind::Success,
-            DispatchResult::success(dispatch, program_id, gas_counter.into()),
+            DispatchResult::success(dispatch, destination_id, gas_counter.into()),
         ));
     }
```

Comment:
Seems that previous `WontExecute` variant was better, but that could be considered to be changed/redesigned later.

Reply:
Yeah, I also liked previous variants more.
It's not always `Err`(-or) occured, but it's okay as tmp solution

Reply:
Completely agree with you guys, however, there are some considerations.

First of all intrinsic `Result` type has a lot of adapters that have to be implemented by ourselves should we use a new custom type.
The second point is that nevertheless `Result::Err` means an erroneous path of execution and should contain an error object there is a result type of the `binary_search` function: it returns an index of insert position in the `Err` variant. Moreover `Err` doesn't mean an error for `binary_search`: it represents the second successful result of the function that searching item is abscent and contains an index would an item sit in a collection. So this is an example of not much idiomatic use of `Result` in std/core.

Based on these applications I decided to switch to `Result` finally.